### PR TITLE
feat!: UI v3

### PR DESCRIPTION
**Pull request by Jabher**
This is PR for tracking all the changes in UIv3. This branch is already publishing as 3.0.0-next.x, but it will take some time to clear things out enough to merge

fixes UI-245

### CODE REVIEW

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  initialTheme,
+  // overrideTheme is mainly used for storybook
+  overrideTheme,
+}) => {
+  const [themeName, setThemeName] = useState<THEME>(
+    initialTheme || DEFAULT_THEME_NAME
+  )
+  const [isAutoDetectedTheme, setIsAutoDetectedTheme] = useState(true)
+
+  console.log('using theme', themeName)
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:46]: Why do we need console.log in prod?

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    initColors()
+    const mql = window.matchMedia('(prefers-color-scheme: dark)')
+    const setTheme = () => {
+      const systemThemeName = mql.matches ? (DARK as THEME) : (LIGHT as THEME)
+      const themeNameCookie = getThemeNameFromCookies()
+      setThemeName(
+        overrideTheme ||
+          themeNameCookie ||
+          initialTheme ||
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:62]: Why does initialTheme go before systemTheme?

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    initialTheme || DEFAULT_THEME_NAME
+  )
+  const [isAutoDetectedTheme, setIsAutoDetectedTheme] = useState(true)
+
+  console.log('using theme', themeName)
+
+  // Noticing browser preferences on hydration
+  // Reacting to changing preferences
+  useEffect(() => {
+    if (typeof window === 'undefined') {
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:51]: useEffect aren't executed during ssr, so it's safe to omit this check

https://nextjs.org/docs/basic-features/data-fetching/client-side#client-side-data-fetching-with-useeffect

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      const systemThemeName = mql.matches ? (DARK as THEME) : (LIGHT as THEME)
+      const themeNameCookie = getThemeNameFromCookies()
+      setThemeName(
+        overrideTheme ||
+          themeNameCookie ||
+          initialTheme ||
+          systemThemeName ||
+          DEFAULT_THEME_NAME
+      )
+      setIsAutoDetectedTheme(
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:66]: Why do we need to disable auto-detection if we prefer themeNameCookie over systemTheme?

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  // Noticing browser preferences on hydration
+  // Reacting to changing preferences
+  useEffect(() => {
+    if (typeof window === 'undefined') {
+      return
+    }
+    initColors()
+    const mql = window.matchMedia('(prefers-color-scheme: dark)')
+    const setTheme = () => {
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:56]: Can we move it out of useEffect to make it a bit smaller?

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        })
+      },
+    }),
+    [themeName]
+  )
+
+  return (
+    <ThemeToggleContext.Provider value={value}>
+      <ThemeProvider
+        isAutoDetectedTheme={isAutoDetectedTheme}
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:92]: why do we need config to enable/disable auto detection?

File: packages/theme/src/provider.tsx [hunk 7]
```diff
@@ -1,23 +1,60 @@
...
+  PropsWithChildren<{ theme: Theme; isAutoDetectedTheme?: boolean }>
+> = ({
+  theme = themeDefault,
+  children,
+  isAutoDetectedTheme = false,
+  ...rest
+}) => {
+  const parentTheme = useThemeToggle()
+  useEffect(() => {
+    if (typeof window === 'undefined') {
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/provider.tsx:29]: We don't need to check for window in useEffects 

https://nextjs.org/docs/basic-features/data-fetching/client-side#client-side-data-fetching-with-useeffect

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+          return newThemeName
+        })
+      },
+    }),
+    [themeName]
+  )
+
+  return (
+    <ThemeToggleContext.Provider value={value}>
+      <ThemeProvider
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/cookie-theme-provider.tsx:91]: Do we need cookie theme provider and regular theme provider? Should they be separated into multiple sub-packages?

File: packages/cookie-theme-toggler/src/hooks.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+import { useContext } from 'react'
+import { ThemeToggleContext, ThemeContext } from './cookie-theme-provider'
+
+export const useThemeToggle = (): ThemeContext => {
```
Comment (kozlovzxc-at-lido) [inline packages/cookie-theme-toggler/src/hooks.ts:4]: What is the benefit of such hook?

File: packages/theme/src/constants.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+export const LIGHT = 'light'
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/constants.ts:1]: Why do we need separate variables? Why not use THEME.light or THEME.dark?

File: packages/theme/src/constants.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+export const DARK = 'dark'
+
+export enum THEME {
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/constants.ts:4]: Couldn't it be const? Const enums has much simpler inner structure

https://www.typescriptlang.org/play?ts=4.4.4#code/KYOwrgtgBAogjFA3gKCmqAbAlgcwBYAuANKugCYCGATgNbIC+yyAxgPYgDOBUwCAvLDgA6bPgJM2nbqEiwATElJpRhEuiiVaDCey48FAmHKGaaQA

File: packages/theme/src/document-head-contents/element-theme-script.tsx [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import React from 'react'
+import { STORAGE_THEME_MANUAL_KEY, ThemeName, VOID_FN } from '../globals'
+import { getTopLevelDomain } from '@lidofinance/utils'
+import {COOKIE_THEME_MANUAL_KEY, COOKIES_THEME_EXPIRES_DAYS} from "../../../cookie-theme-toggler/src/constants";
+
+/**
+ *  this FN should be hermetic and has zero external items - it is inlined via .toString()
+ * */
+/*#__PURE__*/
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/document-head-contents/element-theme-script.tsx:9]: Why does this PURE mean? This function is referring document.cookie, it isn't pure 🤔

File: packages/theme/src/document-head-contents/utils/generate-css-color-variables.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        .replace(/^.*\((.*)\).*$/, '$1')
+        .split(',')
+        .map((n) => parseInt(n))
+        .slice(0, 3)
+    default:
+      return null
+  }
+}
+
+export const generateCssColorVariables = (
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/document-head-contents/utils/generate-css-color-variables.ts:24]: Maybe add docs/examples/tests what this function do? It feels that this function is important

File: packages/theme/src/document-head-contents/element-theme-script.tsx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  match.addEventListener('change', setTheme)
+  return setTheme
+}
+
+export const themeScriptValue =
+  typeof window !== 'undefined' ? themeScriptValueString : () => VOID_FN
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+export let updateGlobalTheme: (theme: string) => void = VOID_FN
+
+export let initGlobalCookieTheme =
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/document-head-contents/element-theme-script.tsx:36]: Maybe describe somewhere some scheme how this works? I think overall idea is simple, but implementation isn't straightforward without much context

File: packages/theme/src/globals.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+export enum ThemeName {
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/globals.ts:1]: What's the difference between Theme above and ThemeName?

File: packages/theme/src/provider.tsx [hunk 16]
```diff
@@ -1,23 +1,60 @@
...
 
-export const ThemeProvider: FC<ThemeProviderProps<Theme>> = (props) => {
-  const { theme = themeDefault, ...rest } = props
+// you need this if you're using themes, so we initiate this automatically
+initColors()
 
-  return <StyledThemeProvider theme={theme} {...rest} />
+const StyledWrapper = styled.div<{ colors?: Record<string, string> }>`
+  display: contents;
+  ${({ colors }) => (colors ? generateCssColorVariables(colors) : null)}
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/provider.tsx:17]: Why do we need it here if we set them in header?

File: packages/theme/src/styles.tsx [hunk 17]
```diff
@@ -0,0 +1,20 @@
+import { FC, PropsWithChildren } from 'react'
+import styled from 'styled-components'
+
+export const Block: FC<PropsWithChildren> = styled.div`
+  font-size: ${({ theme }) => theme.fontSizesMap.xs}px;
+  line-height: 1.5em;
+  padding: ${({ theme }) => theme.spaceMap.xxl}px;
+  margin: ${({ theme }) => theme.spaceMap.md}px 0;
+  border-radius: ${({ theme }) => theme.borderRadiusesMap.xl}px;
+  background: ${({ theme }) => theme.colors.foreground};
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/styles.tsx:10]: Shouldn't it be var(--something)?

File: packages/theme/src/types.ts [hunk 18]
```diff
@@ -1,5 +1,7 @@
 import { themeDefault } from './themes'
 
+export type ThemeName = 'light' | 'dark'
```
Comment (kozlovzxc-at-lido) [inline packages/theme/src/types.ts:3]: One more ThemeName? We have two above, why do we need another one?

File: packages/utils/src/get-top-level-domain.ts [hunk 19]
```diff
@@ -0,0 +1,11 @@
+export const getTopLevelDomain =
+  typeof window === 'undefined'
+    ? () => 'localhost'
```
Comment (kozlovzxc-at-lido) [inline packages/utils/src/get-top-level-domain.ts:3]: Why is it localhost for undefined window?

File: packages/utils/src/get-top-level-domain.ts [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+    : () => {
+        if (document.location.host.indexOf('localhost') === 0) {
```
Comment (kozlovzxc-at-lido) [inline packages/utils/src/get-top-level-domain.ts:5]: It is a bit incorrect:

`'localhost.domain.com'.indexOf('localhost') === 0`

**Comment by kozlovzxc-at-lido** [comment]
Btw, there are few `theme.colors` usages still, we discussed it in dm, but just not to forget

image

File: packages/utils/src/get-top-level-domain.ts (hunk 19 above)
Comment (Jabher) [reply packages/utils/src/get-top-level-domain.ts:3]: we need some default value for easier types coearsing

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx (hunk 2 above)
Comment (Jabher) [reply packages/cookie-theme-toggler/src/cookie-theme-provider.tsx]: to avoid hydration errors

File: packages/theme/src/provider.tsx (hunk 7 above)
Comment (Jabher) [reply packages/theme/src/provider.tsx]: you are right but thinking wrong way - beware that in that manner you are vendor-locking us to next.js

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx (hunk 8 above)
Comment (Jabher) [reply packages/cookie-theme-toggler/src/cookie-theme-provider.tsx]: I've merged them into single /theme repo

File: packages/cookie-theme-toggler/src/hooks.ts (hunk 9 above)
Comment (Jabher) [reply packages/cookie-theme-toggler/src/hooks.ts]: it was not me, but via using approach like this we're making it impossible to use this provider in custom manner

File: packages/theme/src/constants.ts (hunk 11 above)
Comment (Jabher) [reply packages/theme/src/constants.ts]: I'm unsure about swc and const enum, can you please research on that?

File: packages/theme/src/document-head-contents/element-theme-script.tsx (hunk 12 above)
Comment (Jabher) [reply packages/theme/src/document-head-contents/element-theme-script.tsx:8]: this is special syntax for minifiers and transpilers.
basically it allows function to be inlined, but here it is used as a (not fully working but) canary test that should break things if this function will use external logic or variables

File: packages/theme/src/document-head-contents/utils/generate-css-color-variables.ts (hunk 13 above)
Comment (Jabher) [reply packages/theme/src/document-head-contents/utils/generate-css-color-variables.ts]: how the tests should look like? no idea tbh

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx (hunk 2 above)
Comment (solidovic) [reply packages/cookie-theme-toggler/src/cookie-theme-provider.tsx]: outdated

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx (hunk 3 above)
Comment (solidovic) [reply packages/cookie-theme-toggler/src/cookie-theme-provider.tsx]: outdated

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx (hunk 5 above)
Comment (solidovic) [reply packages/cookie-theme-toggler/src/cookie-theme-provider.tsx]: outdated

File: packages/theme/src/styles.tsx (hunk 17 above)
Comment (solidovic) [reply packages/theme/src/styles.tsx]: outdated

File: packages/utils/src/get-top-level-domain.ts (hunk 20 above)
Comment (solidovic) [reply packages/utils/src/get-top-level-domain.ts]: outdated

File: packages/theme/src/types.ts (hunk 18 above)
Comment (solidovic) [reply packages/theme/src/types.ts]: outdated

File: packages/theme/src/provider.tsx (hunk 16 above)
Comment (solidovic) [reply packages/theme/src/provider.tsx]: outdated

File: packages/cookie-theme-toggler/src/cookie-theme-provider.tsx (hunk 4 above)
Comment (solidovic) [reply packages/cookie-theme-toggler/src/cookie-theme-provider.tsx]: outdated

**Comment by solidovic** [comment]
> 

all outdated exclude .storybook

**Review by solidovic — APPROVED** [review]

**Review by kozlovzxc-at-lido — APPROVED** [review]

**Pull request merged**